File: backend/opencode/think_tags.go
package opencode

import (
	"strings"
	"unicode"
)

var (
	thinkOpenTags = []string{
		string([]byte{0x3c, 0x74, 0x68, 0x69, 0x6e, 0x6b, 0x3e}),
		string([]byte{0x3c, 0x72, 0x65, 0x64, 0x61, 0x63, 0x74, 0x65, 0x64, 0x5f, 0x74, 0x68, 0x69, 0x6e, 0x6b, 0x69, 0x6e, 0x67, 0x3e}),
	}
	thinkCloseTags = []string{
		string([]byte{0x3c, 0x2f, 0x74, 0x68, 0x69, 0x6e, 0x6b, 0x3e}),
		string([]byte{0x3c, 0x2f, 0x72, 0x65, 0x64, 0x61, 0x63, 0x74, 0x65, 0x64, 0x5f, 0x74, 0x68, 0x69, 0x6e, 0x6b, 0x69, 0x6e, 0x67, 0x3e}),
	}
)

func maxThinkTagLen() int {
	n := 0
	for _, t := range thinkOpenTags {
		if len(t) > n {
			n = len(t)
		}
	}
	for _, t := range thinkCloseTags {
		if len(t) > n {
			n = len(t)
		}
	}
	return n
}

func findThinkOpen(s string, from int) (idx, tagLen int) {
	lower := strings.ToLower(s[from:])
	best := -1
	bestLen := 0
	for _, tag := range thinkOpenTags {
		if i := strings.Index(lower, strings.ToLower(tag)); i >= 0 && (best < 0 || i < best) {
			best = i
			bestLen = len(tag)
		}
	}
	if best < 0 {
		return -1, 0
	}
	return from + best, bestLen
}

func closeTagForOpen(openTag string) string {
	for i, open := range thinkOpenTags {
		if strings.EqualFold(openTag, open) {
			return thinkCloseTags[i]
		}
	}
	return thinkCloseTags[0]
}

func findThinkCloseFrom(s string, from int, openTag string) (idx, tagLen int) {
	closeTag := closeTagForOpen(openTag)
	lower := strings.ToLower(s[from:])
	if i := strings.Index(lower, strings.ToLower(closeTag)); i >= 0 {
		return from + i, len(closeTag)
	}
	best := -1
	bestLen := 0
	for _, tag := range thinkCloseTags {
		if i := strings.Index(lower, strings.ToLower(tag)); i >= 0 && (best < 0 || i < best) {
			best = i
			bestLen = len(tag)
		}
	}
	if best < 0 {
		return -1, 0
	}
	return from + best, bestLen
}

func skipSpace(s string, from int) int {
	for from < len(s) && unicode.IsSpace(rune(s[from])) {
		from++
	}
	return from
}

func SplitEmbeddedThinking(s string) (text, thinking string) {
	var textParts, thinkParts []string
	i := 0
	for i < len(s) {
		openIdx, openLen := findThinkOpen(s, i)
		if openIdx < 0 {
			textParts = append(textParts, s[i:])
			break
		}
		textParts = append(textParts, s[i:openIdx])
		thinkStart := openIdx + openLen
		openTag := s[openIdx : openIdx+openLen]
		closeIdx, closeLen := findThinkCloseFrom(s, thinkStart, openTag)
		if closeIdx < 0 {
			thinkParts = append(thinkParts, s[thinkStart:])
			break
		}
		thinkParts = append(thinkParts, s[thinkStart:closeIdx])
		i = skipSpace(s, closeIdx+closeLen)
	}
	return strings.Join(textParts, ""), strings.Join(thinkParts, "")
}

func SanitizeEmbeddedThinking(msg *Message) {
	if msg == nil || msg.Role != "assistant" {
		return
	}
	raw := ContentString(*msg)
	if raw == "" {
		return
	}
	text, embedded := SplitEmbeddedThinking(raw)
	if embedded == "" && text == raw {
		return
	}
	if text != "" {
		msg.Content = StringContent(text)
	} else {
		msg.Content = nil
	}
	existing := msg.GetReasoning()
	if embedded != "" {
		if existing != "" {
			existing += embedded
		} else {
			existing = embedded
		}
	}
	if existing != "" {
		msg.ReasoningContent = &existing
		msg.ReasoningDetails = nil
		msg.ReasoningPlain = nil
	}
}

type thinkStreamSplitter struct {
	inThink bool
	openTag string
	carry   strings.Builder
}

func (s *thinkStreamSplitter) feed(chunk string, emitText, emitThink func(string)) {
	data := s.carry.String() + chunk
	s.carry.Reset()
	if data == "" {
		return
	}

	maxHold := maxThinkTagLen() - 1
	i := 0
	for i < len(data) {
		if !s.inThink {
			openIdx, openLen := findThinkOpen(data, i)
			if openIdx < 0 {
				safeEnd := len(data)
				if safeEnd-i > maxHold {
					safeEnd = len(data) - maxHold
				}
				if safeEnd > i {
					emitText(data[i:safeEnd])
				}
				if safeEnd < len(data) {
					s.carry.WriteString(data[safeEnd:])
				}
				return
			}
			if openIdx > i {
				emitText(data[i:openIdx])
			}
			s.inThink = true
			s.openTag = data[openIdx : openIdx+openLen]
			i = openIdx + openLen
			continue
		}

		closeIdx, closeLen := findThinkCloseFrom(data, i, s.openTag)
		if closeIdx < 0 {
			safeEnd := len(data)
			if safeEnd-i > maxHold {
				safeEnd = len(data) - maxHold
			}
			if safeEnd > i {
				emitThink(data[i:safeEnd])
			}
			if safeEnd < len(data) {
				s.carry.WriteString(data[safeEnd:])
			}
			return
		}
		if closeIdx > i {
			emitThink(data[i:closeIdx])
		}
		s.inThink = false
		s.openTag = ""
		i = skipSpace(data, closeIdx+closeLen)
	}
}

func (s *thinkStreamSplitter) flush(emitText, emitThink func(string)) {
	if s.carry.Len() == 0 {
		return
	}
	rest := s.carry.String()
	s.carry.Reset()
	if s.inThink {
		emitThink(rest)
	} else {
		emitText(rest)
	}
}

File: backend/opencode/think_tags.ts
// PORT: backend/opencode/think_tags.go

import { get_reasoning, string_content, content_string, type message } from "./types";

const think_open_tags = [
  "<think>",
  "<redacted_thinking>"
];

const think_close_tags = [
  "</think>",
  "</redacted_thinking>"
];

export const max_think_tag_len = (): number => {
  let n = 0;
  for (const t of think_open_tags) {
    if (t.length > n) {
      n = t.length;
    }
  }
  for (const t of think_close_tags) {
    if (t.length > n) {
      n = t.length;
    }
  }
  return n;
};

export const find_think_open = (s: string, from: number): [number, number] => {
  const lower = s.slice(from).toLowerCase();
  let best = -1;
  let best_len = 0;
  for (const tag of think_open_tags) {
    const i = lower.indexOf(tag.toLowerCase());
    if (i >= 0 && (best < 0 || i < best)) {
      best = i;
      best_len = tag.length;
    }
  }
  if (best < 0) {
    return [-1, 0];
  }
  return [from + best, best_len];
};

export const close_tag_for_open = (open_tag: string): string => {
  const lower_open = open_tag.toLowerCase();
  for (let i = 0; i < think_open_tags.length; i++) {
    if (think_open_tags[i].toLowerCase() === lower_open) {
      return think_close_tags[i];
    }
  }
  return think_close_tags[0];
};

export const find_think_close_from = (s: string, from: number, open_tag: string): [number, number] => {
  const close_tag = close_tag_for_open(open_tag);
  const lower = s.slice(from).toLowerCase();
  const first_idx = lower.indexOf(close_tag.toLowerCase());
  if (first_idx >= 0) {
    return [from + first_idx, close_tag.length];
  }
  let best = -1;
  let best_len = 0;
  for (const tag of think_close_tags) {
    const i = lower.indexOf(tag.toLowerCase());
    if (i >= 0 && (best < 0 || i < best)) {
      best = i;
      best_len = tag.length;
    }
  }
  if (best < 0) {
    return [-1, 0];
  }
  return [from + best, best_len];
};

export const skip_space = (s: string, from: number): number => {
  while (from < s.length && /\s/.test(s[from])) {
    from++;
  }
  return from;
};

export const split_embedded_thinking = (s: string): [string, string] => {
  const text_parts: string[] = [];
  const think_parts: string[] = [];
  let i = 0;
  while (i < s.length) {
    const [open_idx, open_len] = find_think_open(s, i);
    if (open_idx < 0) {
      text_parts.push(s.slice(i));
      break;
    }
    text_parts.push(s.slice(i, open_idx));
    const think_start = open_idx + open_len;
    const open_tag = s.slice(open_idx, open_idx + open_len);
    const [close_idx, close_len] = find_think_close_from(s, think_start, open_tag);
    if (close_idx < 0) {
      think_parts.push(s.slice(think_start));
      break;
    }
    think_parts.push(s.slice(think_start, close_idx));
    i = skip_space(s, close_idx + close_len);
  }
  return [text_parts.join(""), think_parts.join("")];
};

export const sanitize_embedded_thinking = (msg: message | null | undefined): void => {
  if (!msg || msg.role !== "assistant") {
    return;
  }
  const raw = content_string(msg);
  if (raw === "") {
    return;
  }
  const [text, embedded] = split_embedded_thinking(raw);
  if (embedded === "" && text === raw) {
    return;
  }
  if (text !== "") {
    msg.content = string_content(text);
  } else {
    msg.content = new Uint8Array();
  }
  let existing = get_reasoning(msg);
  if (embedded !== "") {
    if (existing !== "") {
      existing += embedded;
    } else {
      existing = embedded;
    }
  }
  if (existing !== "") {
    msg.reasoning_content = existing;
    delete msg.reasoning_details;
    delete msg.reasoning;
  }
};

export class think_stream_splitter {
  in_think = false;
  open_tag = "";
  carry = "";

  feed(chunk: string, emit_text: (text: string) => void, emit_think: (text: string) => void): void {
    const data = this.carry + chunk;
    this.carry = "";
    if (data === "") {
      return;
    }

    const max_hold = max_think_tag_len() - 1;
    let i = 0;
    while (i < data.length) {
      if (!this.in_think) {
        const [open_idx, open_len] = find_think_open(data, i);
        if (open_idx < 0) {
          let safe_end = data.length;
          if (safe_end - i > max_hold) {
            safe_end = data.length - max_hold;
          }
          if (safe_end > i) {
            emit_text(data.slice(i, safe_end));
          }
          if (safe_end < data.length) {
            this.carry += data.slice(safe_end);
          }
          return;
        }
        if (open_idx > i) {
          emit_text(data.slice(i, open_idx));
        }
        this.in_think = true;
        this.open_tag = data.slice(open_idx, open_idx + open_len);
        i = open_idx + open_len;
        continue;
      }

      const [close_idx, close_len] = find_think_close_from(data, i, this.open_tag);
      if (close_idx < 0) {
        let safe_end = data.length;
        if (safe_end - i > max_hold) {
          safe_end = data.length - max_hold;
        }
        if (safe_end > i) {
          emit_think(data.slice(i, safe_end));
        }
        if (safe_end < data.length) {
          this.carry += data.slice(safe_end);
        }
        return;
      }
      if (close_idx > i) {
        emit_think(data.slice(i, close_idx));
      }
      this.in_think = false;
      this.open_tag = "";
      i = skip_space(data, close_idx + close_len);
    }
  }

  flush(emit_text: (text: string) => void, emit_think: (text: string) => void): void {
    if (this.carry.length === 0) {
      return;
    }
    const rest = this.carry;
    this.carry = "";
    if (this.in_think) {
      emit_think(rest);
    } else {
      emit_text(rest);
    }
  }
}

/*
PORT STATUS
source path: backend/opencode/think_tags.go
source lines: 216
draft lines: 182
confidence: high
status: phase_b_compile
*/
